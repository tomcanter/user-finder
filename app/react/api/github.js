import reqwest from 'reqwest';
import parseLinkHeader from 'utils/parseLinkHeader';
import path from 'path';

const API_ROOT_URL = 'https://api.github.com';

function gitHubAPI(url) {
  var req = reqwest({
    url: path.join(API_ROOT_URL, url),
    type: 'json'
  })
    .then(data => {
      return {
        data,
        pagination: parseLinkHeader(req.request.getResponseHeader('Link'))
      }
    });

  return req;
}

export default gitHubAPI;