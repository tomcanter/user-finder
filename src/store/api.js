import GitHub from 'github-api';

const api = new GitHub({
  token: process.env.USER_FINDER_OAUTH,
});

export default api;
