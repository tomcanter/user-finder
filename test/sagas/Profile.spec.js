import {
  call,
  put,
} from 'redux-saga/effects';
import api from 'store/api';
import {
  getProfile,
  getRepos,
  getFollowers,
} from 'store/sagas/Profile';

describe('Saga: getProfile', () => {

  describe('when no cached data exists', () => {
    it('should call the API and put a success action with the profile data', () => {
      const generator = getProfile({username: 'tomcanter'});
      const profile = generator.next().value;
      expect(profile).toEqual(
        call(api.getProfile, 'tomcanter')
      );
    });

    it('should put actions for repos and followers', () => {
      const response = {
        data: {
          login: 'tomcanter',
          followers_url: 'followers.net',
          repos_url: 'repos.net',
        },
      };
      const generator = getProfile({username: 'tomcanter'});
      generator.next();
      generator.next(response);

      const repoRequestAction = generator.next().value;
      expect(repoRequestAction).toEqual(
        put({type: 'REPOS_REQUEST', url: 'repos.net'})
      );

      const followerRequestAction = generator.next().value;
      expect(followerRequestAction).toEqual(
        put({type: 'FOLLOWERS_REQUEST', url: 'followers.net'})
      );
    });
  });

});

describe('Saga: getFollowers', () => {

  describe('when requesting followers', () => {
    it('should call the API with the url and parameters', () => {
      const generator = getFollowers({url: 'followers.net'});
      const followers = generator.next().value;
      expect(followers.CALL).toMatchSnapshot();
    });

    it('should put a success action with the normalized data', () => {
      const generator = getFollowers({url: 'followers.net'});
      const response = {
        data: [
          {id: 1, login: 'test'},
          {id: 2, login: 'hello'},
        ],
      };
      generator.next();
      const normalized = generator.next(response).value;
      expect(normalized.PUT).toMatchSnapshot();
    });
  });

  describe('when a request to the API fails', () => {
    it('should call a failure action', () => {
      const generator = getFollowers({url: 'followers.net'});
      const apiError = new Error('it went wrong');
      apiError.response = 'test';

      generator.next();
      const errorAction = generator.throw(apiError).value;
      expect(errorAction.PUT).toMatchSnapshot();
    });
  });

});

describe('Saga: getRepos', () => {

  describe('when requesting repos', () => {
    it('should call the API with the url and parameters', () => {
      const generator = getRepos({url: 'repos.net'});
      const repos = generator.next().value;
      expect(repos.CALL).toMatchSnapshot();
    });

    it('should put a success action with the normalized data', () => {
      const generator = getRepos({url: 'repos.net'});
      const response = {
        data: [
          {id: 1, name: 'test'},
          {id: 2, name: 'hello'},
        ],
      };
      generator.next();
      const normalized = generator.next(response).value;
      expect(normalized.PUT).toMatchSnapshot();
    });
  });

  describe('when a request to the API fails', () => {
    it('should call a failure action', () => {
      const generator = getRepos({url: 'repos.net'});
      const apiError = new Error('it went wrong');
      apiError.response = 'test';

      generator.next();
      const errorAction = generator.throw(apiError).value;
      expect(errorAction.PUT).toMatchSnapshot();
    });
  });

});
