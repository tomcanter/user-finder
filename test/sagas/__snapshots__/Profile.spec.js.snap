// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Saga: getFollowers when a request to the API fails should call a failure action 1`] = `
Object {
  "action": Object {
    "message": "it went wrong",
    "response": "test",
    "type": "FOLLOWERS_FAILURE",
  },
  "channel": null,
}
`;

exports[`Saga: getFollowers when requesting followers should call the API with the url and parameters 1`] = `
Object {
  "args": Array [
    "followers.net",
    Object {
      "params": Object {
        "per_page": 8,
      },
    },
  ],
  "context": null,
  "fn": [Function],
}
`;

exports[`Saga: getFollowers when requesting followers should put a success action with the normalized data 1`] = `
Object {
  "action": Object {
    "entities": Object {
      "users": Object {
        "1": Object {
          "id": 1,
          "login": "test",
        },
        "2": Object {
          "id": 2,
          "login": "hello",
        },
      },
    },
    "result": Array [
      1,
      2,
    ],
    "type": "FOLLOWERS_SUCCESS",
  },
  "channel": null,
}
`;

exports[`Saga: getRepos when a request to the API fails should call a failure action 1`] = `
Object {
  "action": Object {
    "message": "it went wrong",
    "response": "test",
    "type": "REPOS_FAILURE",
  },
  "channel": null,
}
`;

exports[`Saga: getRepos when requesting repos should call the API with the url and parameters 1`] = `
Object {
  "args": Array [
    "repos.net",
    Object {
      "params": Object {
        "per_page": 10,
        "sort": "pushed",
      },
    },
  ],
  "context": null,
  "fn": [Function],
}
`;

exports[`Saga: getRepos when requesting repos should put a success action with the normalized data 1`] = `
Object {
  "action": Object {
    "entities": Object {
      "repos": Object {
        "1": Object {
          "id": 1,
          "name": "test",
        },
        "2": Object {
          "id": 2,
          "name": "hello",
        },
      },
    },
    "result": Array [
      1,
      2,
    ],
    "type": "REPOS_SUCCESS",
  },
  "channel": null,
}
`;
